crates/ruff_linter/src/rules/ruff/rules/operation_on_closed_io.rs

```diff
@@ -0,0 +1,171 @@
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::AnyNodeRef;
+use ruff_python_semantic::{Binding, BindingKind, NodeId, SemanticModel};
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for usages of IO operation methods of context variables
+/// outside of the original `with` statement.
+///
+/// ## Why is this bad?
+/// Such operations will raise `ValueError: I/O operation on closed file` at runtime.
+///
+/// ## Example
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w"):
+///     f.write("")
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w") as f:
+///     f.write("")
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct OperationOnClosedIO;
+
+impl Violation for OperationOnClosedIO {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "IO operation performed on closed IO object".to_string()
+    }
+}
+
+/// RUF061
+pub(crate) fn operation_on_closed_io(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    if !matches!(&binding.kind, BindingKind::WithItemVar) {
+        return None;
+    }
+
+    let semantic = checker.semantic();
+    let with = binding.statement(semantic)?.as_with_stmt()?;
```

Comment:
I think we should validate that `f` is a file and not just any other context variable (assert that it was created using `open`)